### PR TITLE
Fix branch compiler and execution

### PR DESCRIPTION
**Pull request by EngHabu**
# TL;DR
Fixes branch compiler and branch execution.

## The compiler had two wrong assumptions:
- All nodes in a branch nodes will have the same inputs and those inputs must be bound in the parent node. This is an unnecessary restriction. Removed.
- Nodes inside a branch nodes will have globally unique IDs. This is not necessarily true. Added code to build unique node ids.

## Execution runtime had a regression:
A prior PR removed the persistence of datadir in the workflow CRD. Branch handler relies on that persistence to find the selected branch's outputs location. The fix is to reverse that setting and update the selected branch's output location to the parent's... This way once the execution of the branch is done, the outputs are already available in the parent node's location.

## Type
 - [ ] Bug Fix
 - [X] Feature
 - [ ] Plugin

## Are all requirements met?

 - [X] Code completed
 - [X] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [x] Any pending items have an associated Issue

## Tracking Issue
https://github.com/lyft/flyte/issues/602

### CODE REVIEW

File: pkg/apis/flyteworkflow/v1alpha1/workflow.go [hunk 1]
```diff
@@ -157,7 +157,7 @@ func (in *Inputs) UnmarshalJSON(b []byte) error {
 
 func (in *Inputs) MarshalJSON() ([]byte, error) {
 	if in == nil {
-		return []byte{}, nil
+		return json.Marshal(nil)
```
Comment (kumare3) [inline pkg/apis/flyteworkflow/v1alpha1/workflow.go:160]: i guess we could make this a constant and re-use it?

File: pkg/compiler/validators/utils.go [hunk 2]
```diff
@@ -115,6 +117,26 @@ func withVariableName(param *core.Variable) (newParam *core.Variable, ok bool) {
 	return
 }
 
+func JoinVariableMapsUniqueKeys(m1, m2 map[string]*core.Variable) (map[string]*core.Variable, error) {
```
Comment (kumare3) [inline pkg/compiler/validators/utils.go:120]: do you imply union of maps?

**Comment by kumare3** [comment]
This looks like a big change, a lot of null checks and fixes.

File: pkg/compiler/validators/branch.go [hunk 3]
```diff
@@ -31,55 +31,63 @@ func validateBranchInterface(w c.WorkflowBuilder, node c.NodeBuilder, errs error
...
+		if otherCase.ThenNode == nil {
 			errs.Collect(errors.NewValueRequiredErr(node.GetId(), "IfElse.Case.ThenNode"))
 			continue
 		}
 
-		n := w.NewNodeBuilder(block.ThenNode)
+		cases = append(cases, otherCase.ThenNode)
+	}
+
+	if elseNode := node.GetBranchNode().IfElse.GetElseNode(); elseNode != nil {
```
Comment (kumare3) [inline pkg/compiler/validators/branch.go:59]: you missed out on the else.Fail case

**Comment by kumare3** [comment]
Looks good to me, except for one comment. To be honest, its very hard to verify the entire logic :)

File: pkg/compiler/validators/node.go [hunk 4]
```diff
@@ -81,6 +86,8 @@ func ValidateBranchNode(w c.WorkflowBuilder, n c.NodeBuilder, requireParamType b
 	if elseNode := n.GetBranchNode().IfElse.GetElseNode(); elseNode != nil {
 		wrapperNode := w.NewNodeBuilder(elseNode)
 		subNodes = append(subNodes, wrapperNode)
+	} else if defaultError := n.GetBranchNode().IfElse.GetError(); defaultError == nil {
```
Comment (kumare3) [inline pkg/compiler/validators/node.go]: do you not handle the  error today?

Comment (EngHabu) [reply pkg/compiler/validators/node.go]: nope...

**Review by kumare3 — APPROVED** [review]
